File: webengage_flutter/ios/Classes/WebEngagePlugin.m
#import <WebEngage/WebEngage.h>

#import "WebEngagePlugin.h"
#import "WebEngageConstants.h"

NSString *const kMessagingPresentationOptionsUserDefaults =
@"webengage_flutter_presentation_options";

@implementation WebEngagePlugin {
    FlutterMethodChannel *_channel;
}

- (instancetype)initWithFlutterMethodChannel:(FlutterMethodChannel *)channel andFlutterPluginRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    self = [super init];
    if (self) {
        _channel = channel;
    }
    return self;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    NSLog(@"[WebEngagePlugin] registerWithRegistrar");
    FlutterMethodChannel *channel = [FlutterMethodChannel methodChannelWithName:WEBENGAGE_PLUGIN binaryMessenger:[registrar messenger]];
    id instance = [[WebEngagePlugin alloc] initWithFlutterMethodChannel:channel andFlutterPluginRegistrar:registrar];
    [WebEngage sharedInstance].pushNotificationDelegate = instance;
    [registrar addMethodCallDelegate:instance channel:channel];
    [registrar addApplicationDelegate:instance];
}

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    NSLog(@"[WebEngagePlugin] didFinishLaunchingWithOptions");
    return [[WebEngage sharedInstance] application:application didFinishLaunchingWithOptions:launchOptions notificationDelegate:self];
}

- (BOOL)application:(UIApplication *)application continueUserActivity:(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray *))restorationHandler {
    NSLog(@"[WebEngagePlugin] continueUserActivity");
    [[[WebEngage sharedInstance] deeplinkManager] getAndTrackDeeplink:userActivity.webpageURL callbackBlock:^(id location){
        [self trackDeeplinkURLCallback:location];
    }];
    return YES;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([METHOD_NAME_REQUEST_PERMISSION isEqualToString:call.method]) {
        [self messagingRequestPermission:call withResult:result];
    } else if ([METHOD_NAME_SET_FOREGROUND_NOTIFICATION_PRESENTATION_OPTIONS isEqualToString:call.method]) {
        [self messagingSetForegroundNotificationPresentationOptions:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_LOGIN isEqualToString:call.method]) {
        [self userLogin:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_LOGOUT isEqualToString:call.method]) {
        [self userLogout:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_FIRST_NAME isEqualToString:call.method]) {
        [self setUserFirstName:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_LAST_NAME isEqualToString:call.method]) {
        [self setUserLastName:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_EMAIL isEqualToString:call.method]) {
        [self setUserEmail:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_HASHED_EMAIL isEqualToString:call.method]) {
        [self setUserHashedEmail:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_PHONE isEqualToString:call.method]) {
        [self setUserPhone:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_HASHED_PHONE isEqualToString:call.method]) {
        [self setUserHashedPhone:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_COMPANY isEqualToString:call.method]) {
        [self setUserCompany:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_BIRTHDATE isEqualToString:call.method]) {
        [self setUserBirthDate:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_GENDER isEqualToString:call.method]) {
        [self setUserGender:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_OPT_IN isEqualToString:call.method]) {
        [self setUserOptIn:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_LOCATION isEqualToString:call.method]) {
        [self setUserLocation:call withResult:result];
    } else if ([METHOD_NAME_TRACK_EVENT isEqualToString:call.method]) {
        [self trackEvent:call withResult:result];
    } else if ([METHOD_NAME_TRACK_SCREEN isEqualToString:call.method]) {
        [self trackScreen:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_ATTRIBUTE isEqualToString:call.method]) {
        [self setUserAttribute:call withResult:result];
    } else if ([METHOD_NAME_SET_USER_MAP_ATTRIBUTE isEqualToString:call.method]) {
        [self setUserAttributes:call withResult:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void) messagingRequestPermission:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
    
    UNAuthorizationOptions options = UNAuthorizationOptionNone;
    
    if ([call.arguments[@"alert"] isEqual:@(YES)]) {
        options |= UNAuthorizationOptionAlert;
    }
    
    if ([call.arguments[@"badge"] isEqual:@(YES)]) {
        options |= UNAuthorizationOptionBadge;
    }
    
    if ([call.arguments[@"sound"] isEqual:@(YES)]) {
        options |= UNAuthorizationOptionSound;
    }
    
    if ([call.arguments[@"provisional"] isEqual:@(YES)]) {
        if (@available(iOS 12.0, *)) {
            options |= UNAuthorizationOptionProvisional;
        }
    }
    
    if ([call.arguments[@"announcement"] isEqual:@(YES)]) {
        if (@available(iOS 13.0, *)) {
            options |= UNAuthorizationOptionAnnouncement;
        }
    }
    
    if ([call.arguments[@"carPlay"] isEqual:@(YES)]) {
        options |= UNAuthorizationOptionCarPlay;
    }
    
    if ([call.arguments[@"criticalAlert"] isEqual:@(YES)]) {
        if (@available(iOS 12.0, *)) {
            options |= UNAuthorizationOptionCriticalAlert;
        }
    }
    
    id handler = ^(BOOL granted, NSError *_Nullable error) {
        if (error != nil) {
            result(@(NO));
        } else {
            [center getNotificationSettingsWithCompletionHandler:^(
                                                                   UNNotificationSettings *_Nonnull settings) {
                result(@(YES));
            }];
        }
    };
    
    [center requestAuthorizationWithOptions:options completionHandler:handler];
}

- (void)messagingSetForegroundNotificationPresentationOptions:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSMutableDictionary *persistedOptions = [NSMutableDictionary dictionary];
    if ([call.arguments[@"alert"] isEqual:@(YES)]) {
        persistedOptions[@"alert"] = @YES;
    }
    if ([call.arguments[@"badge"] isEqual:@(YES)]) {
        persistedOptions[@"badge"] = @YES;
    }
    if ([call.arguments[@"sound"] isEqual:@(YES)]) {
        persistedOptions[@"sound"] = @YES;
    }
    
    [[NSUserDefaults standardUserDefaults] setObject:persistedOptions
                                              forKey:kMessagingPresentationOptionsUserDefaults];
    result(nil);
}


- (void) userLogin:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * userId = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser login:userId];
    result(nil);
}

- (void) userLogout:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser logout];
    result(nil);
}

- (void) setUserFirstName:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * firstName = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setFirstName:firstName];
    result(nil);
}

- (void) setUserLastName:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * lastName = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setLastName:lastName];
    result(nil);
}

- (void) setUserEmail:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * email = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setEmail:email];
    result(nil);
}

- (void) setUserHashedEmail:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * hashedEmail = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setHashedEmail:hashedEmail];
    result(nil);
}

- (void) setUserPhone:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * phone = call.arguments;
    WEGUser* weUser = [WebEngage sharedInstance].user;
    [weUser setPhone:phone];
    result(nil);
}

- (void) setUserHashedPhone:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * hashedPhone = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setHashedPhone:hashedPhone];
    result(nil);
}

- (void) setUserCompany:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * company = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setCompany:company];
    result(nil);
}

- (void) setUserBirthDate:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * birthDate = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setBirthDateString:birthDate];
    result(nil);
}

- (void) setUserGender:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * gender = call.arguments;
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setGender:gender];
    result(nil);
}

- (void) setUserOptIn:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * channel = call.arguments[CHANNEL];
    NSLocale* locale = [NSLocale localeWithLocaleIdentifier:@"en_US"];
    NSString* ch = [channel lowercaseStringWithLocale:locale];
    
    BOOL status;
    
    if ([call.arguments[OPTIN] isEqual:@(YES)]) {
        status = YES;
    } else {
        status = NO;
    }
    
    WEGUser * weUser = [WebEngage sharedInstance].user;
    
    if ([ch isEqualToString:PUSH]) {
        [weUser setOptInStatusForChannel:WEGEngagementChannelPush status:status];
        result(nil);
    } else if ([ch isEqualToString:SMS]) {
        [weUser setOptInStatusForChannel:WEGEngagementChannelSMS status:status];
        result(nil);
    } else if ([ch isEqualToString:EMAIL]) {
        [weUser setOptInStatusForChannel:WEGEngagementChannelEmail status:status];
        result(nil);
    } else if ([ch isEqualToString:IN_APP]) {
        [weUser setOptInStatusForChannel:WEGEngagementChannelInApp status:status];
        result(nil);
    } else if ([ch isEqualToString:WHATSAPP]) {
        [weUser setOptInStatusForChannel:WEGEngagementChannelWhatsapp status:status];
        result(nil);
    } else {
        NSString * msg = [NSString stringWithFormat:@"Invalid channel: %@. Must be one of [push, sms, email, in_app, whatsapp].", ch];
        result([FlutterError errorWithCode:@"WebEngagePlugin" message:msg details:nil]);
    }
}

- (void) setUserLocation:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSNumber * lat = call.arguments[LAT];
    NSNumber * lng = call.arguments[LNG];
    WEGUser * weUser = [WebEngage sharedInstance].user;
    [weUser setUserLocationWithLatitude:lat andLongitude:lng];
    result(nil);
}

- (void) trackEvent:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * eventName = call.arguments[EVENT_NAME];
    NSDictionary * attributes = call.arguments[ATTRIBUTES];
    id<WEGAnalytics> weAnalytics = [WebEngage sharedInstance].analytics;
    if (attributes && ![attributes isKindOfClass:[NSNull class]]) {
        [weAnalytics trackEventWithName:eventName andValue:attributes];
    }
    else{
        [weAnalytics trackEventWithName:eventName];
    }
    result(nil);
}

- (void) trackScreen:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString * screenName = call.arguments[SCREEN_NAME];
    NSDictionary * screenData = call.arguments[SCREEN_DATA];
    id<WEGAnalytics> weAnalytics = [WebEngage sharedInstance].analytics;
    if (screenData && ![screenData isKindOfClass:[NSNull class]]) {
        [weAnalytics navigatingToScreenWithName:screenName andData:screenData];
    }
    else{
        [weAnalytics navigatingToScreenWithName:screenName];
    }
    result(nil);
}

- (void) setUserAttribute:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString* attributeName = call.arguments[ATTRIBUTE_NAME];
    id value = call.arguments[ATTRIBUTES];
    WEGUser * weUser = [WebEngage sharedInstance].user;
    if ([value isKindOfClass:[NSString class]]) {
        if ([value length] == 24) {
            NSDate * date = [self getDate:value];
            if (date != nil) {
                [weUser setAttribute:attributeName withDateValue:date];
            } else {
                [weUser setAttribute:attributeName withStringValue:value];
            }
        } else {
            [weUser setAttribute:attributeName withStringValue:value];
        }
    }
    else if ([value isKindOfClass:[NSNumber class]]) {
        [weUser setAttribute:attributeName withValue:value];
    }
    else if ([value isKindOfClass:[NSArray class]]) {
        [weUser setAttribute:attributeName withArrayValue:value];
    }
    else if ([value isKindOfClass:[NSDate class]]) {
        [weUser setAttribute:attributeName withDateValue:value];
    }
    result(nil);
}

- (void) setUserAttributes:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString* attributeName = call.arguments[ATTRIBUTE_NAME];
    id value = call.arguments[ATTRIBUTES];
    if ([value isKindOfClass:[NSDictionary class]]) {
        [[WebEngage sharedInstance].user setAttribute:attributeName withDictionaryValue:value];
    }
    result(nil);
    
}

- (NSDate *)getDate:(NSString *)strValue {
    NSDateFormatter * dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"];
    [dateFormatter setTimeZone:[NSTimeZone timeZoneWithAbbreviation:@"UTC"]];
    NSDate * date = [dateFormatter dateFromString:strValue];
    return date;
}

- (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler {
    if ([notification.request.content.userInfo[@"source"] isEqual:@"webengage"]) {
        NSLog(@"[WebEngagePlugin] willPresentNotification");
        NSMutableDictionary *customData = [[NSMutableDictionary alloc] init];
        for (NSDictionary* pair in notification.request.content.userInfo[@"customData"])
        {
            [customData setValue:pair[@"value"] forKey:pair[@"key"]];
        }
        NSDictionary *payload = @{
            TITLE: notification.request.content.title,
            MESSAGE: notification.request.content.body,
            STATE: NOTIFICATION_STATE_SHOWN,
            DATA: customData,
        };
        [_channel invokeMethod:METHOD_NAME_ON_PUSH arguments:payload];
    }
    
    UNNotificationPresentationOptions presentationOptions = UNNotificationPresentationOptionNone;
    NSDictionary *persistedOptions = [[NSUserDefaults standardUserDefaults]
                                      dictionaryForKey:kMessagingPresentationOptionsUserDefaults];
    if (persistedOptions != nil) {
        if ([persistedOptions[@"alert"] isEqual:@(YES)]) {
            presentationOptions |= UNNotificationPresentationOptionAlert;
        }
        if ([persistedOptions[@"badge"] isEqual:@(YES)]) {
            presentationOptions |= UNNotificationPresentationOptionBadge;
        }
        if ([persistedOptions[@"sound"] isEqual:@(YES)]) {
            presentationOptions |= UNNotificationPresentationOptionSound;
        }
    }
    completionHandler(presentationOptions);
}

- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void(^)(void))completionHandler {
    if ([response.actionIdentifier isEqualToString:UNNotificationDismissActionIdentifier] &&
        [response.notification.request.content.userInfo[@"source"] isEqual:@"webengage"]) {
        NSLog(@"[WebEngagePlugin] didReceiveNotificationResponse");
        NSMutableDictionary *customData = [[NSMutableDictionary alloc] init];
        for (NSDictionary* pair in response.notification.request.content.userInfo[@"customData"])
        {
            [customData setValue:pair[@"value"] forKey:pair[@"key"]];
        }
        NSDictionary *payload = @{
            TITLE: response.notification.request.content.title,
            MESSAGE: response.notification.request.content.body,
            STATE: NOTIFICATION_STATE_DISMISSED,
            DATA: customData,
        };
        [_channel invokeMethod:METHOD_NAME_ON_PUSH arguments:payload];
    }
    completionHandler();
}

- (void)WEGHandleDeeplink:(NSString *)deeplink userData:(NSDictionary *)data {
    NSLog(@"[WebEngagePlugin] WEGHandleDeeplink");
    NSMutableDictionary *customData = [[NSMutableDictionary alloc] init];
    for (NSDictionary* pair in data[@"customData"])
    {
        [customData setValue:pair[@"value"] forKey:pair[@"key"]];
    }
    NSDictionary *payload = @{
        TITLE: data[@"aps"][@"alert"][@"title"],
        MESSAGE: data[@"aps"][@"alert"][@"body"],
        STATE: NOTIFICATION_STATE_CLICKED,
        DATA: customData,
        LINK: deeplink,
    };
    [_channel invokeMethod:METHOD_NAME_ON_PUSH arguments:payload];
}

- (NSDictionary *)notificationPrepared:(NSDictionary<NSString *,id> *)inAppNotificationData shouldStop:(BOOL *)stopRendering {
    NSLog(@"[WebEngagePlugin] notificationPrepared");
    NSDictionary *payload = @{
        STATE:IN_APP_STATE_PREPARED,
        DATA:inAppNotificationData,
    };
    [_channel invokeMethod:METHOD_NAME_ON_IN_APP arguments:payload];
    return inAppNotificationData;
}

- (void)notificationShown:(NSDictionary<NSString *,id> *)inAppNotificationData {
    NSLog(@"[WebEngagePlugin] notificationShown");
    NSDictionary *payload = @{
        STATE:IN_APP_STATE_SHOWN,
        DATA:inAppNotificationData,
    };
    [_channel invokeMethod:METHOD_NAME_ON_IN_APP arguments:payload];
}

- (void)notificationDismissed:(NSDictionary<NSString *,id> *)inAppNotificationData {
    NSLog(@"[WebEngagePlugin] notificationDismissed");
    NSDictionary *payload = @{
        STATE:IN_APP_STATE_DISMISSED,
        DATA:inAppNotificationData,
    };
    [_channel invokeMethod:METHOD_NAME_ON_IN_APP arguments:payload];
}

- (void)notification:(NSMutableDictionary<NSString *,id> *)inAppNotificationData clickedWithAction:(NSString *)actionId {
    NSLog(@"[WebEngagePlugin] notification clickedWithAction");
    NSMutableDictionary *payload = [[NSMutableDictionary alloc] init];
    [payload setValue:IN_APP_STATE_CLICKED forKey:STATE];
    [payload setValue:inAppNotificationData forKey:DATA];
    if (actionId != (id)[NSNull null] && actionId.length > 0 ) {
        for (NSDictionary* action in inAppNotificationData[@"actions"])
        {
            if ([actionId isEqualToString:action[@"actionEId"]]) {
                [payload setValue:action[@"actionLink"] forKey:LINK];
            }
        }
    }
    
    [_channel invokeMethod:METHOD_NAME_ON_IN_APP arguments:payload];
}

- (void)trackDeeplinkURLCallback:(NSString *)redirectLocationURL {
    NSLog(@"[WebEngagePlugin] trackDeeplinkURLCallback");
    [_channel invokeMethod:METHOD_NAME_ON_DEEP_LINK arguments:redirectLocationURL];
}

@end

File: webengage_flutter/android/src/main/java/com/webengage/webengage_plugin/WebEngagePlugin.java
package com.webengage.webengage_plugin;

import android.app.Activity;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.content.pm.PackageManager;
import android.os.Bundle;

import com.google.firebase.messaging.FirebaseMessaging;
import com.webengage.sdk.android.Channel;
import com.webengage.sdk.android.LocationTrackingStrategy;
import com.webengage.sdk.android.WebEngage;
import com.webengage.sdk.android.WebEngageConfig;
import com.webengage.sdk.android.utils.Gender;

import java.util.Collections;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import static com.webengage.webengage_plugin.Constants.ARGS.*;
import static com.webengage.webengage_plugin.Constants.MethodName.*;
import static com.webengage.webengage_plugin.Constants.WEBENGAGE_PLUGIN;
import static com.webengage.webengage_plugin.Constants.TAG;

public class WebEngagePlugin implements FlutterPlugin, MethodCallHandler, ActivityAware {
    private static MethodChannel channel;
    private static final Map<String, Map<String, Object>> messageQueue =
            Collections.synchronizedMap(new LinkedHashMap<>());
    public static boolean isInitialised = false;
    public static boolean isAttachedToEngine = false;

    private Activity activity;

    public static String getPackageName(Context context) {
        try {
            return context.getPackageManager().getPackageInfo(
                    context.getPackageName(), 0).packageName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String getMetadata(Context context, String key) {
        try {
            Bundle metaData = context.getPackageManager()
                    .getApplicationInfo(context.getPackageName(),
                            PackageManager.GET_META_DATA).metaData;
            return metaData.get(key).toString();
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public void onAttachedToEngine(FlutterPluginBinding flutterPluginBinding) {
        this.channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), WEBENGAGE_PLUGIN);
        channel.setMethodCallHandler(this);
        this.isAttachedToEngine = true;
        if (!isInitialised) {
            initialize(flutterPluginBinding.getApplicationContext());
        }
        FirebaseMessaging.getInstance().getToken().addOnSuccessListener(token -> WebEngage.get().setRegistrationID(token));
        synchronized (messageQueue) {
            // Handle all the messages received before the Dart isolate was
            // initialized, then clear the queue.
            for (Map.Entry<String, Map<String, Object>> entry :
                    messageQueue.entrySet()) {
                sendCallback(entry.getKey(), entry.getValue());
            }
            messageQueue.clear();
        }
    }

    @Override
    public void onDetachedFromEngine(FlutterPluginBinding binding) {
        this.isAttachedToEngine = false;
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        switch (call.method) {
            case METHOD_NAME_SET_USER_LOGIN:
                userLogin(call, result);
                break;

            case METHOD_NAME_SET_USER_LOGOUT:
                userLogout(call, result);
                break;

            case METHOD_NAME_SET_USER_FIRST_NAME:
                setUserFirstName(call, result);
                break;

            case METHOD_NAME_SET_USER_LAST_NAME:
                setUserLastName(call, result);
                break;

            case METHOD_NAME_SET_USER_EMAIL:
                setUserEmail(call, result);
                break;

            case METHOD_NAME_SET_USER_HASHED_EMAIL:
                setUserHashedEmail(call, result);
                break;

            case METHOD_NAME_SET_USER_PHONE:
                setUserPhone(call, result);
                break;

            case METHOD_NAME_SET_USER_HASHED_PHONE:
                setUserHashedPhone(call, result);
                break;

            case METHOD_NAME_SET_USER_COMPANY:
                setUserCompany(call, result);
                break;

            case METHOD_NAME_SET_USER_BIRTHDATE:
                setUserBirthDate(call, result);
                break;

            case METHOD_NAME_SET_USER_GENDER:
                setUserGender(call, result);
                break;

            case METHOD_NAME_SET_USER_OPT_IN:
                setUserOptIn(call, result);
                break;

            case METHOD_NAME_SET_USER_LOCATION:
                setUserLocation(call, result);
                break;

            case METHOD_NAME_TRACK_EVENT:
                trackEvent(call, result);
                break;

            case METHOD_NAME_TRACK_SCREEN:
                trackScreen(call, result);
                break;

            case METHOD_NAME_SET_USER_ATTRIBUTE:
                setUserAttribute(call, result);
                break;

            case METHOD_NAME_SET_USER_MAP_ATTRIBUTE:
                setUserMapAttribute(call, result);
                break;

            default:
                result.notImplemented();
        }
    }

    public static void initialize(Context context) {
        WebEngage.registerPushNotificationCallback(new WebEngageNotificationCallbacks());
        WebEngage.registerInAppNotificationCallback(new WebEngageInAppCallbacks());
        WebEngageConfig config =
                new WebEngageConfig.Builder()
                        .setWebEngageKey(getMetadata(context, "com.webengage.sdk.android.key"))
                        .build();
        WebEngage.engage(context, config);
        isInitialised = true;
        WebEngageBackgroundService.onInitialized();
    }

    private void userLogin(MethodCall call, Result result) {
        String userId = call.arguments();
        WebEngage.get().user().login(userId);
        result.success(null);
    }

    private void userLogout(MethodCall call, Result result) {
        WebEngage.get().user().logout();
        result.success(null);
    }

    private void setUserFirstName(MethodCall call, Result result) {
        String firstName = call.arguments();
        WebEngage.get().user().setFirstName(firstName);
        result.success(null);
    }

    private void setUserLastName(MethodCall call, Result result) {
        String lastName = call.arguments();
        WebEngage.get().user().setLastName(lastName);
        result.success(null);
    }

    private void setUserEmail(MethodCall call, Result result) {
        String email = call.arguments();
        WebEngage.get().user().setEmail(email);
        result.success(null);
    }

    private void setUserHashedEmail(MethodCall call, Result result) {
        String hashedEmail = call.arguments();
        WebEngage.get().user().setHashedEmail(hashedEmail);
        result.success(null);
    }

    private void setUserPhone(MethodCall call, Result result) {
        String phone = call.arguments();
        WebEngage.get().user().setPhoneNumber(phone);
        result.success(null);
    }

    private void setUserHashedPhone(MethodCall call, Result result) {
        String hashedPhone = call.arguments();
        WebEngage.get().user().setHashedPhoneNumber(hashedPhone);
        result.success(null);
    }

    private void setUserCompany(MethodCall call, Result result) {
        String company = call.arguments();
        WebEngage.get().user().setCompany(company);
        result.success(null);
    }

    private void setUserBirthDate(MethodCall call, Result result) {
        String birthDate = call.arguments();
        WebEngage.get().user().setBirthDate(birthDate);
        result.success(null);
    }

    private void setUserGender(MethodCall call, Result result) {
        String gender = call.arguments();
        if (MALE.equalsIgnoreCase(gender)) {
            WebEngage.get().user().setGender(Gender.MALE);
        } else if (FEMALE.equalsIgnoreCase(gender)) {
            WebEngage.get().user().setGender(Gender.FEMALE);
        } else if (OTHER.equalsIgnoreCase(gender)) {
            WebEngage.get().user().setGender(Gender.OTHER);
        }
        result.success(null);
    }

    private void setUserOptIn(MethodCall call, Result result) {
        String channel = call.argument(CHANNEL);
        boolean status = call.argument(OPTIN);
        if (PUSH.equalsIgnoreCase(channel)) {
            WebEngage.get().user().setOptIn(Channel.PUSH, status);
        } else if (SMS.equalsIgnoreCase(channel)) {
            WebEngage.get().user().setOptIn(Channel.SMS, status);
        } else if (EMAIL.equalsIgnoreCase(channel)) {
            WebEngage.get().user().setOptIn(Channel.EMAIL, status);
        } else if (IN_APP.equalsIgnoreCase(channel)) {
            WebEngage.get().user().setOptIn(Channel.IN_APP, status);
        } else if (WHATSAPP.equalsIgnoreCase(channel)) {
            WebEngage.get().user().setOptIn(Channel.WHATSAPP, status);
        } else {
            result.error(TAG, "Invalid channel: " + channel + ". Must be one of [push, sms, email, in_app, whatsapp].", null);
        }
        result.success(null);
    }

    private void setUserLocation(MethodCall call, Result result) {
        double lat = call.argument(LAT);
        double lng = call.argument(LNG);
        WebEngage.get().user().setLocation(lat, lng);
        result.success(null);
    }

    private void setUserMapAttribute(MethodCall call, Result result) {
        Map<String, ? extends Object> attributes = call.argument(ATTRIBUTES);
        WebEngage.get().user().setAttributes(attributes);
        result.success(null);
    }

    private void setUserAttribute(MethodCall call, Result result) {
        if (call.argument(ATTRIBUTES) instanceof String) {
            String attributeName = call.argument(ATTRIBUTE_NAME);
            String attributes = call.argument(ATTRIBUTES);
            WebEngage.get().user().setAttribute(attributeName, attributes);
        } else if (call.argument(ATTRIBUTES) instanceof Integer) {
            String attributeName = call.argument(ATTRIBUTE_NAME);
            int attributes = call.argument(ATTRIBUTES);
            WebEngage.get().user().setAttribute(attributeName, attributes);
        } else if (call.argument(ATTRIBUTES) instanceof Double || call.argument(ATTRIBUTES) instanceof Float) {
            String attributeName = call.argument(ATTRIBUTE_NAME);
            double attributes = call.argument(ATTRIBUTES);
            WebEngage.get().user().setAttribute(attributeName, attributes);
        } else if (call.argument(ATTRIBUTES) instanceof Date) {
            String attributeName = call.argument(ATTRIBUTE_NAME);
            Date attributes = call.argument(ATTRIBUTES);
            WebEngage.get().user().setAttribute(attributeName, attributes);
        } else if (call.argument(ATTRIBUTES) instanceof List) {
            String attributeName = call.argument(ATTRIBUTE_NAME);
            List<? extends Object> attributes = call.argument(ATTRIBUTES);
            WebEngage.get().user().setAttribute(attributeName, attributes);
        } else if (call.argument(ATTRIBUTES) instanceof Boolean) {
            String attributeName = call.argument(ATTRIBUTE_NAME);
            Boolean attributes = call.argument(ATTRIBUTES);
            WebEngage.get().user().setAttribute(attributeName, attributes);
        } else {
            Log.d(TAG, "No other type supported");
        }
        result.success(null);
    }

    private void trackEvent(MethodCall call, Result result) {
        String eventName = call.argument(EVENT_NAME);
        Map<String, Object> attributes = call.argument(ATTRIBUTES);
        WebEngage.get().analytics().track(eventName, attributes);
        result.success(null);
    }

    private void trackScreen(MethodCall call, Result result) {
        String screenName = call.argument(SCREEN_NAME);
        Map<String, Object> screenData = call.argument(SCREEN_DATA);
        if (screenData == null) {
            WebEngage.get().analytics().screenNavigated(screenName);
        } else {
            WebEngage.get().analytics().screenNavigated(screenName, screenData);
        }
        result.success(null);
    }

    static void sendOrQueueCallback(String methodName, Map<String, Object> message) {
        if (isAttachedToEngine) {
            sendCallback(methodName, message);
        } else {
            messageQueue.put(methodName, message);
        }
    }

    static void sendCallback(final String methodName, final Map<String, Object> message) {
        new Handler(Looper.getMainLooper()).post(() -> channel.invokeMethod(methodName, message));
    }

    @Override
    public void onAttachedToActivity(ActivityPluginBinding binding) {
        this.activity = binding.getActivity();
        WebEngage.get().analytics().start(activity);
    }

    @Override
    public void onReattachedToActivityForConfigChanges(ActivityPluginBinding binding) {
        onAttachedToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        WebEngage.get().analytics().stop(activity);
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity();
    }
}
